### PR TITLE
8271883: Math CopySign optimization for x86

### PR DESCRIPTION
**Pull request by mgkwill**
Intrinsic for Math.copySign is disabled on x86_64.

We can improve on generated c2 instructions for float and double, and this change adds optimized intrinsics for float and double Math.copySign.

### **Math.copySign(double)**
_From:_
  0x00007f7d606e5dac:   vmovq  %xmm1,%r10
  0x00007f7d606e5db1:   vmovq  %xmm0,%r11
  0x00007f7d606e5db6:   movabs $0x7fffffffffffffff,%r8
  0x00007f7d606e5dc0:   and    %r8,%r11
  0x00007f7d606e5dc3:   movabs $0x8000000000000000,%r8
  0x00007f7d606e5dcd:   and    %r8,%r10
  0x00007f7d606e5dd0:   or     %r11,%r10
  0x00007f7d606e5dd3:   vmovq  %r10,%xmm0                   

_To:_
  0x00007fc3c14c63ac:   movabs $0x7fffffffffffffff,%r10
  0x00007fc3c14c63b6:   vmovq  %r10,%xmm2
  0x00007fc3c14c63bb:   vpternlogq $0xe4,%xmm2,%xmm1,%xmm0

### **Math.copySign(float)**
_From:_
  0x00007ff8886e60ac:   vmovd  %xmm1,%r11d
  0x00007ff8886e60b1:   vmovd  %xmm0,%r10d
  0x00007ff8886e60b6:   and    $0x80000000,%r11d
  0x00007ff8886e60bd:   and    $0x7fffffff,%r10d
  0x00007ff8886e60c4:   or     %r10d,%r11d
  0x00007ff8886e60c7:   vmovd  %r11d,%xmm0
_To:_
  0x00007fc7d94c63ac:   mov    $0x7fffffff,%r10d
  0x00007fc7d94c63b2:   vmovd  %r10d,%xmm3
  0x00007fc7d94c63b7:   vpternlogd $0xe4,%xmm3,%xmm1,%xmm0

#### **BEFORE**
Signum._5_copySignFloatTest       avgt    5  2.442 ? 0.024  ns/op
Signum._7_copySignDoubleTest      avgt    5  2.400 ? 0.033  ns/op

#### **PATCH**
Signum._5_copySignFloatTest       avgt    5  2.029 ? 0.011  ns/op
Signum._7_copySignDoubleTest      avgt    5  2.029 ? 0.024  ns/op

#### **JTREG that covers this case:**
test/hotspot/jtreg/compiler/intrinsics/math/TestSignumIntrinsic.java

Signed-off-by: Marcus G K Williams <marcus.williams@intel.com>

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8271883](https://bugs.openjdk.java.net/browse/JDK-8271883): Math CopySign optimization for x86

### Reviewers
 * [Jatin Bhateja](https://openjdk.java.net/census#jbhateja) (@jatin-bhateja - Committer)
 * [Sandhya Viswanathan](https://openjdk.java.net/census#sviswanathan) (@sviswa7 - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5005/head:pull/5005` \
`$ git checkout pull/5005`

Update a local copy of the PR: \
`$ git checkout pull/5005` \
`$ git pull https://git.openjdk.java.net/jdk pull/5005/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5005`

View PR using the GUI difftool: \
`$ git pr show -t 5005`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5005.diff">https://git.openjdk.java.net/jdk/pull/5005.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86.ad:1692]: This should be part of match_rule_supported() method and not match_rule_supported_vector().
Also should return false for non LP_64.

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86.ad:5814]: Please add the tmp registers to the format, something like below:
format %{ "CopySignF $dst, $src\t! using $tmp1 and $tmp2 as TEMP" %}

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86.ad:5810]: rRegL tmp2 should be rRegI tmp2 for copySignF.

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86.ad:5828]: Please add the tmp1 and tmp2 registers to the format.

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86.ad:5842]: Please add the tmp1 and tmp2 registers to the format.

File: src/hotspot/cpu/x86/x86.ad
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/x86.ad:5846]: No harm in adding a comment describing logic expression encoded by E4.

File: src/hotspot/cpu/x86/x86.ad
Comment (jatin-bhateja) [inline src/hotspot/cpu/x86/x86.ad:5823]: Graph shape created for this will always pass zero operand as an immediate argument. Do you see any issue if you just keep one pattern (copySignD_imm) ?

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86.ad:5835]: You could use movdl() instead of adding a new vmod instruction.

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86.ad:5848]: You could use movq() instead of adding a new vmovq instruction.

File: src/hotspot/cpu/x86/x86.ad
Comment (sviswa7) [inline src/hotspot/cpu/x86/x86.ad:5813]: Good to add then C (xmm2) is set to 0x7FFFFFFF.

File: src/hotspot/cpu/x86/x86.ad
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/x86.ad]: @mgkwill , haven't seen your response on this.

Comment (mgkwill) [reply src/hotspot/cpu/x86/x86.ad]: @jatin-bhateja, @sviswa7 told me that you and her discussed it and decided that removing copySignD_reg was appropriate as it is not needed. I've removed copySignD_reg.

Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/x86.ad]: Thanks! patch looks good otherwise.

**Review by jatin-bhateja — APPROVED** [review]

**Review by sviswa7 — APPROVED** [review]

**Comment by sviswa7** [comment]
Could we please get second review from a reviewer? Also need help running this through Oracle's testing.

**Comment by mgkwill** [comment]
@vnkozlov any chance you have time to run this through Oracle Testing?

I tested tier1-3 on cascade lake Linux x86_64.

**Comment by vnkozlov** [comment]
I started our testing.

**Comment by vnkozlov** [comment]
Testing passed clean.

**Comment by sviswa7** [comment]
@vnkozlov Thanks a lot Vladimir.  Jatin and I have reviewed the patch. Could we integrate or should we wait for another review.

**Comment by mgkwill** [comment]
> Testing passed clean.

Thanks @vnkozlov!

**Review by vnkozlov — APPROVED** [review]
Yes, you can push it,

**Pull request closed without merging**
